File: apps/desktop/src-tauri/src/selection/positioning.rs
use super::SelectionSnapshot;
use crate::capture::MonitorInfo;

const GAP: i32 = 8;
const MARGIN: i32 = 12;
const TOOLBAR_HEIGHT: i32 = 44;
const TOOLBAR_WIDTH: i32 = 360;

fn clamp(value: i32, min: i32, max: i32) -> i32 {
    value.max(min).min(max)
}

/// Screen bounds for a compact always-on-top toolbar window (physical pixels).
pub fn toolbar_window_bounds(snapshot: &SelectionSnapshot, monitor: &MonitorInfo) -> (i32, i32, u32, u32) {
    let scale = if monitor.scale_factor > 0.0 {
        monitor.scale_factor
    } else {
        1.0
    };

    let sel_left = (snapshot.x - monitor.x) as f64 / scale;
    let sel_top = (snapshot.y - monitor.y) as f64 / scale;
    let sel_bottom = sel_top + (snapshot.height as f64 / scale).max(1.0);

    let vh = (monitor.height as f64 / scale) as i32;
    let vw = (monitor.width as f64 / scale) as i32;

    let mut top = sel_bottom.round() as i32 + GAP;
    if top + TOOLBAR_HEIGHT > vh - MARGIN {
        top = sel_top.round() as i32 - TOOLBAR_HEIGHT - GAP;
    }
    top = clamp(top, MARGIN, vh - TOOLBAR_HEIGHT - MARGIN);

    let left = clamp(sel_left.round() as i32, MARGIN, vw - TOOLBAR_WIDTH - MARGIN);

    let screen_x = monitor.x + (left as f64 * scale).round() as i32;
    let screen_y = monitor.y + (top as f64 * scale).round() as i32;
    let width = (TOOLBAR_WIDTH as f64 * scale).round().max(120.0) as u32;
    let height = (TOOLBAR_HEIGHT as f64 * scale).round().max(32.0) as u32;

    (screen_x, screen_y, width, height)
}

File: apps/desktop/src/selection/positioning.ts
import type { CSSProperties } from 'react';
import type { OverlayMonitor, SelectionRect } from './types';

import {
  POPUP_DEFAULT_HEIGHT,
  POPUP_DEFAULT_WIDTH,
} from './popup-hooks';

const GAP = 8;
const MARGIN = 12;
const TOOLBAR_HEIGHT = 44;
const POPUP_WIDTH = POPUP_DEFAULT_WIDTH;
const POPUP_EST_HEIGHT = POPUP_DEFAULT_HEIGHT;

function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(value, max));
}

export function toolbarPositionStyle(rect: SelectionRect): CSSProperties {
  const vh = window.innerHeight;
  const vw = window.innerWidth;

  let top = rect.bottom + GAP;
  if (top + TOOLBAR_HEIGHT > vh - MARGIN) {
    top = rect.top - TOOLBAR_HEIGHT - GAP;
  }
  top = clamp(top, MARGIN, vh - TOOLBAR_HEIGHT - MARGIN);

  const left = clamp(rect.left, MARGIN, vw - 320 - MARGIN);

  return {
    position: 'fixed',
    top: `${top}px`,
    left: `${left}px`,
    pointerEvents: 'auto',
    zIndex: 2147483647,
  };
}

export function getPopupPosition(rect: SelectionRect): { top: number; left: number } {
  const vh = window.innerHeight;
  const vw = window.innerWidth;
  const width = Math.min(POPUP_WIDTH, vw - MARGIN * 2);

  let top = rect.bottom + GAP + TOOLBAR_HEIGHT;
  if (top + POPUP_EST_HEIGHT > vh - MARGIN) {
    top = rect.top - POPUP_EST_HEIGHT - GAP;
  }
  top = clamp(top, MARGIN, vh - POPUP_EST_HEIGHT - MARGIN);

  const left = clamp(rect.left, MARGIN, vw - width - MARGIN);

  return { top, left };
}

export function popupScreenBounds(
  rect: SelectionRect,
  monitor: OverlayMonitor,
  position?: { top: number; left: number },
): { x: number; y: number; width: number; height: number } {
  const scale = monitor.scaleFactor > 0 ? monitor.scaleFactor : 1;
  const pos = position ?? getPopupPosition(rect);
  const cssWidth = Math.min(POPUP_WIDTH, monitor.width / scale - MARGIN * 2);
  const cssHeight = Math.min(POPUP_DEFAULT_HEIGHT, monitor.height / scale - pos.top - MARGIN);

  return {
    x: monitor.x + Math.round(pos.left * scale),
    y: monitor.y + Math.round(pos.top * scale),
    width: Math.round(cssWidth * scale),
    height: Math.round(cssHeight * scale),
  };
}

/**
 * Native window bounds (physical px) for a toolbar of a measured CSS size,
 * anchored under the selection. Mirrors `toolbar_window_bounds` in Rust, but
 * uses the real rendered width so no button gets clipped.
 */
export function toolbarScreenBounds(
  rect: SelectionRect,
  monitor: OverlayMonitor,
  cssWidth: number,
  cssHeight: number,
): { x: number; y: number; width: number; height: number } {
  const scale = monitor.scaleFactor > 0 ? monitor.scaleFactor : 1;
  const vw = monitor.width / scale;
  const vh = monitor.height / scale;

  const width = Math.min(cssWidth, vw - MARGIN * 2);
  const height = Math.min(cssHeight, vh - MARGIN * 2);

  let top = rect.bottom + GAP;
  if (top + height > vh - MARGIN) {
    top = rect.top - height - GAP;
  }
  top = clamp(top, MARGIN, Math.max(MARGIN, vh - height - MARGIN));

  const left = clamp(rect.left, MARGIN, Math.max(MARGIN, vw - width - MARGIN));

  return {
    x: monitor.x + Math.round(left * scale),
    y: monitor.y + Math.round(top * scale),
    width: Math.max(Math.round(width * scale), 1),
    height: Math.max(Math.round(height * scale), 1),
  };
}

export function monitorOverlayBounds(monitor: OverlayMonitor): {
  x: number;
  y: number;
  width: number;
  height: number;
} {
  return {
    x: monitor.x,
    y: monitor.y,
    width: monitor.width,
    height: monitor.height,
  };
}

export function popupPositionStyle(
  rect: SelectionRect,
  position?: { top: number; left: number },
): CSSProperties {
  const vh = window.innerHeight;
  const vw = window.innerWidth;
  const width = Math.min(POPUP_WIDTH, vw - MARGIN * 2);
  const { top, left } = position ?? getPopupPosition(rect);

  return {
    position: 'fixed',
    top: `${top}px`,
    left: `${left}px`,
    width: `${width}px`,
    maxHeight: `${vh - top - MARGIN}px`,
    pointerEvents: 'auto',
    zIndex: 2147483647,
  };
}
